From df1a548a2f81d329e8fe7e6c0df0b41906c4354b Mon Sep 17 00:00:00 2001
From: Sonny Sasaka <sonnysasaka@chromium.org>
Date: Mon, 11 Mar 2019 13:02:37 -0700
Subject: [PATCH] CHROMIUM: Bluetooth: Honor BT_ENFORCE_CLASSIC_SECURITY in
 more places

We want to be able to configure some boards to not have this classic
security enforcement at all. So, this configuration should be honored at
all places where the enforcement happens.

BUG=chromium:929763
TEST=Tested on caroline

Change-Id: I2ee9e9501b4a54a1e8a936af65e91e261f6f0347
Signed-off-by: Sonny Sasaka <sonnysasaka@chromium.org>
Reviewed-on: https://chromium-review.googlesource.com/1521812
---
 net/bluetooth/hci_event.c | 22 ++++++++++++++++++++--
 1 file changed, 20 insertions(+), 2 deletions(-)

diff --git a/net/bluetooth/hci_event.c b/net/bluetooth/hci_event.c
index 843c09a2af05..d769e24b32c1 100644
--- a/net/bluetooth/hci_event.c
+++ b/net/bluetooth/hci_event.c
@@ -2850,12 +2850,15 @@ static void read_enc_key_size_complete(struct hci_dev *hdev, u8 status,
 	 * enough.
 	 */
 	if (rp->status) {
-		WARN(1, "Read Encryption Key Size command failed, chip may not support this");
 		bt_dev_err(hdev, "failed to read key size for handle %u",
 			   handle);
+		conn->enc_key_size = HCI_LINK_KEY_SIZE;
+#ifdef CONFIG_BT_ENFORCE_CLASSIC_SECURITY
+		WARN(1, "Read Encryption Key Size command failed, chip may not support this");
 		hci_disconnect(conn, HCI_ERROR_REMOTE_USER_TERM);
 		hci_conn_drop(conn);
 		goto unlock;
+#endif
 	} else {
 		conn->enc_key_size = rp->key_size;
 	}
@@ -2957,22 +2960,37 @@ static void hci_encrypt_change_evt(struct hci_dev *hdev, struct sk_buff *skb)
 		struct hci_cp_read_enc_key_size cp;
 		struct hci_request req;
 
+		/* Only send HCI_Read_Encryption_Key_Size if the
+		 * controller really supports it. If it doesn't, assume
+		 * the default size (16).
+		 */
+		if (!(hdev->commands[20] & 0x10)) {
+			conn->enc_key_size = HCI_LINK_KEY_SIZE;
+			goto notify;
+		}
+
 		hci_req_init(&req, hdev);
 
 		cp.handle = cpu_to_le16(conn->handle);
 		hci_req_add(&req, HCI_OP_READ_ENC_KEY_SIZE, sizeof(cp), &cp);
 
 		if (hci_req_run_skb(&req, read_enc_key_size_complete)) {
-			WARN(1, "Failed sending HCI Read Encryption Key Size, chip may not support this");
 			bt_dev_err(hdev, "sending read key size failed");
+			conn->enc_key_size = HCI_LINK_KEY_SIZE;
+#ifdef CONFIG_BT_ENFORCE_CLASSIC_SECURITY
+			WARN(1, "Failed sending HCI Read Encryption Key Size, chip may not support this");
 			hci_disconnect(conn, HCI_ERROR_REMOTE_USER_TERM);
 			hci_conn_drop(conn);
 			goto unlock;
+#else
+			goto notify;
+#endif
 		}
 
 		goto unlock;
 	}
 
+notify:
 	if (conn->state == BT_CONFIG) {
 		if (!ev->status)
 			conn->state = BT_CONNECTED;
-- 
2.23.0.rc1.153.gdeed80330f-goog

